### PR TITLE
Refactoring SourceLoader (#96 #89 #137) Importing project from external sources (#126 #127)

### PR DESCRIPTION
**Pull request by cibo94**
For now only tests are implemented.

### CODE REVIEW

**Comment by jsilhan** [comment]
Was it indended to export `tests/project/tars/sample.tar.gz` and `tests/project/tars/sample.tar.xz` too?

File: rpg/source_loader.py [hunk 1]
```diff
@@ -44,20 +43,26 @@ def _get_compression_method(self, name):
         return None
 
     def create_archive(self, path, extracted_dir, compression=TAR_GZIP):
-        name = os.path.basename(str(path)) + ".tar.gz"
+        name = str(path) + ".tar.gz"
         mode = self._tar_compression_mode.get(
             self._tar_compression_type.get(compression))
-        with tar.open(name, 'w:' + mode) as tarfile:
-            tarfile.add(str(extracted_dir), arcname=os.path.basename(str(path)))
+        if os.path.isdir(str(extracted_dir)) or \
```
Comment (jsilhan) [inline rpg/source_loader.py]: you can do `if extracted_dir.exists():` instead

File: rpg/source_loader.py [hunk 2]
```diff
@@ -44,20 +43,26 @@ def _get_compression_method(self, name):
...
+                os.path.isfile(str(extracted_dir)):
+            with tar.open(name, 'w:' + mode) as tarfile:
+                tarfile.add(
+                    str(extracted_dir),
+                    arcname=os.path.basename(str(path)))
```
Comment (jsilhan) [inline rpg/source_loader.py]: you can do this cleaner using Pathlib

File: tests/test_source_loader.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    FNF_MD5 = "d41d8cd98f00b204e9800998ecf8427e"
+
+    def setUp(self):
+        self._source_loader = SourceLoader()
+        self._tar = False
+        self._tar_dir = self.test_project_dir
+        self._tar_gz = self.test_project_dir / "archives" / "sample.tar.gz"
+        self._tar_xz = self.test_project_dir / "archives" / "sample.tar.xz"
+        self._tar_temp = "/var/tmp/rpg_test/"
+        self._tar_extracted = self._tar_temp + "sample"
```
Comment (jsilhan) [inline tests/test_source_loader.py]: use Pathlib

**Comment by jsilhan** [comment]
add into test file check, please, i.e.:
```python
...
...
self.assertExistInDir(["file1", "file2"], _tar_extracted)
```

**Comment by cibo94** [comment]
Almost done please check it out. Do we need more tests?

**Comment by cibo94** [comment]
Importing
=======

I have implemented 2 methods that downloads archives from external sources. Tests are really simple (just to show that this works), so if it will be needed I may implement more of them.

Synopsis:
-------------
```python
SourceLoader.download_git_repo(file_name, git_user, git_repo [, callback] [, branch])
SourceLoader.download_archive(url, file_name [, callback] [, arch_type])
```

Both of them are class methods.

First one have 3 parameters that are required and 2 more optional. 
Branch default value is ```'master'```. Callback is there to report progress of download (default value is ```None```).

Second one download archive from whatever url. It has Callback parameter too to report progress. Arch_type is type of archive with default value ```.tar.gz```.

File: tests/test_source_loader.py [hunk 4]
```diff
@@ -88,24 +84,37 @@ def test_tar_xz_extract(self):
...
 
     def test_dir_source_loader(self):
         self._source_loader.load_sources(
-            str(self.test_project_dir),
+            self.test_project_dir,
             self._tar_extracted)
         self.assertEqual(
             self.md5Dir(self._tar_extracted),
-            self.md5Dir(str(self.test_project_dir)))
+            self.md5Dir(self.test_project_dir))
```
Comment (jsilhan) [inline tests/test_source_loader.py]: Can you put `self.assertExistInDir` there too?

File: rpg/source_loader.py [hunk 5]
```diff
@@ -1,117 +1,175 @@
...
-                                   "rar", "7z", "tgz", "tbz2"] \
-            and not arch_t.group(2) in ["gz", "xz", "lz",
-                                        "bz2", "Z", "lzma"]:
+        if not arch_t.group(1) in cls._compressions \
+                and not arch_t.group(2) in cls._compressions[arch_t.group(1)]:
             return None
         return (arch_t.group(1), arch_t.group(2))
+
+    @classmethod
+    def download_git_repo(cls, file_name, user, repo,
```
Comment (jsilhan) [inline rpg/source_loader.py]: can it take just URL of root project dir on github? I.e.: `https://github.com/rh-lab-q/rpg/`?

File: rpg/source_loader.py [hunk 6]
```diff
@@ -1,117 +1,175 @@
...
+                          callback=None, branch='master'):
+        """ Downloads archive from github """
+
+        cls.download_archive("https://github.com/" + user + "/" +
+                             repo + "/tarball/" + branch,
+                             file_name, callback)
+
+    @classmethod
+    def download_archive(cls, url, file_name,
```
Comment (jsilhan) [inline rpg/source_loader.py]: Could it be just `download_archive(cls, url, callback=None)`? Is there any reason why to separate arch_type from file_name?

**Comment by jsilhan** [comment]
Great you're back on the track and you've fixed the critical loading issue.

File: rpg/source_loader.py (hunk 5 above)
Comment (jsilhan) [inline rpg/source_loader.py]: The use case is to have in GUI just one column that could be path to local file, url or github url.

**Comment by cibo94** [comment]
All done!

**Comment by cibo94** [comment]
All done! tests are passing

**Comment by jsilhan** [comment]
f968c65 merged, thanks

**Comment by jsilhan** [comment]
Merged, thanks.

**Pull request closed without merging**